<a href="https://colab.research.google.com/github/arnold8968/Fordham-NLP/blob/master/NLPQuestion1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
print('hello')

hello


### Question 1

Remotely read all poem files from the storm server into a table, which has five columns -
Author, Title, Tags, Body, and Link. Each poem is a row of the table.

Connect to datasets using google drive. Import the text one by one.

In [ ]:
from google.colab import drive
drive.mount("/content/gdrive")

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [ ]:
import pandas as pd
import numpy as np
import os
import re

filePath = '/content/gdrive/Shared drives/CISC6210NLPFall20/LoveOutput/'
filelist = os.listdir(filePath)

In [ ]:
filelist[: 3]

['OnWorld-Making-.txt', 'OperaBouffe-ByP.txt', 'Orpheus-ByAFMor.txt']

In [ ]:
len(filelist)

2235

There are 2235 files in the datasets. We need to choose a method which can save the RAM and robust. So I close the file after import the data.

In [ ]:
df = pd.DataFrame(columns =  ['author', 'title', 'tags', 'body', 'link', 'file_name'])

In [ ]:
for file in filelist:
    test =[]
    with open(filePath + file, encoding='windows-1254') as file:
        for line in file:
            if line != '\n':
                print(line)
                test.append(line.strip('\n'))
    file.close()
    
    # Some of the file may not have the body or link. We can use the try and except to avoid exception 
    try:
        title= test[0].split('By ')[0]
    except:
        title = np.nan
    
    try:
        author= test[0].split('By ')[1]
    except:
        author = np.nan
        
    try:
        tags = test[1].split(';')
    except:
        tags = np.nan

    try:
        body = test[2]
        body = body.replace('<br><br><br>', '[P]').replace('<br><br>', '[P]').replace('<br>', '[L]').replace('\xa0', '')
        body = re.sub(r'<.*>', '', body)
        body = ' '.join(body.split())
    except:
        body = np.nan
        
    try:
        link = test[3].split(': ')[1]
    except:
        link = np.nan
    
    df_length = len(df)
    df.loc[df_length] = [author, title, tags, body, link, file.name]

Streaming output truncated to the last 5000 lines.
He wishes his Beloved were Dead By William Butler Yeats

Love ;Break-ups & Vexed Love ;Heartache & Loss ;Realistic & Complicated ;Relationships ;Men & Women ;Rhymed Stanza

Were you but lying cold and dead, <br>And lights were paling out of the West, <br>You would come hither, and bend your head, <br>And I would lay my head on your breast; <br>And you would murmur tender words, <br>Forgiving me, because you were dead: <br>Nor would you rise and hasten away, <br>Though you have the will of wild birds, <br>But know your hair was bound and wound <br>About the stars and moon and sun: <br>O would, beloved, that you lay <br>Under the dock-leaves in the ground, <br>While lights were paling one by one.<br><br>

original link: https://www.poetryfoundation.org/poems/57305/he-wishes-his-beloved-were-dead
Heat By Denis Johnson

Disappointment & Failure ;Love ;Desire ;Sonnet ;Simile

Here in the electric dusk your naked lover<br> tips the glass hig

In [ ]:
df.head()

author  ...                                          file_name
0    Nomi Stone  ...  /content/gdrive/Shared drives/CISC6210NLPFall2...
1  Philip Gross  ...  /content/gdrive/Shared drives/CISC6210NLPFall2...
2  A. F. Moritz  ...  /content/gdrive/Shared drives/CISC6210NLPFall2...
3  R. S. Thomas  ...  /content/gdrive/Shared drives/CISC6210NLPFall2...
4    Lynda Hull  ...  /content/gdrive/Shared drives/CISC6210NLPFall2...

[5 rows x 6 columns]

--total number of poems stored

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2235 entries, 0 to 2234
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author     2235 non-null   object
 1   title      2235 non-null   object
 2   tags       2235 non-null   object
 3   body       2042 non-null   object
 4   link       2042 non-null   object
 5   file_name  2235 non-null   object
dtypes: object(6)
memory usage: 122.2+ KB


--total number of authors

In [ ]:
len(df['author'].value_counts())

1044

--sort authors by the amount of their poems collected in the table, and show top 20 authors.

In [ ]:
df['author'].value_counts().head(20)

William Shakespeare           52
John Donne                    29
Edmund Spenser                23
Sir  Thomas Wyatt             19
Robert Browning               19
Anonymous                     19
John Keats                    17
Algernon Charles Swinburne    15
Thomas Campion                15
Alfred, Lord Tennyson         13
Robert Burns                  13
George Meredith               13
Brenda Shaughnessy            12
William Butler Yeats          12
Christina Rossetti            12
Aphra Behn                    12
Andrew Marvell                12
Carl Phillips                 12
Amy Lowell                    11
D. A. Powell                  11
Name: author, dtype: int64




The total number of files are 2235 but there only have 2042 body. Let's see why this happened



In [ ]:
df[df.isnull().any(axis=1)]['file_name']

5       /content/gdrive/Shared drives/CISC6210NLPFall2...
19      /content/gdrive/Shared drives/CISC6210NLPFall2...
29      /content/gdrive/Shared drives/CISC6210NLPFall2...
47      /content/gdrive/Shared drives/CISC6210NLPFall2...
51      /content/gdrive/Shared drives/CISC6210NLPFall2...
                              ...                        
2176    /content/gdrive/Shared drives/CISC6210NLPFall2...
2179    /content/gdrive/Shared drives/CISC6210NLPFall2...
2209    /content/gdrive/Shared drives/CISC6210NLPFall2...
2229    /content/gdrive/Shared drives/CISC6210NLPFall2...
2232    /content/gdrive/Shared drives/CISC6210NLPFall2...
Name: file_name, Length: 193, dtype: object

In [ ]:
df.loc[19]['file_name']

'/content/gdrive/Shared drives/CISC6210NLPFall20/LoveOutput/OurPeopleI-ByCM.txt'

In [ ]:
with open(df.loc[19]['file_name'], encoding='windows-1254') as file:
    for line in file:
        if line != '\n':
            print(line)

Our People I By CM Burroughs

Living ;Life Choices ;The Body ;Love ;Realistic & Complicated ;Activities ;Sports & Outdoor Activities ;Relationships ;Family & Ancestors ;Social Commentaries ;Race & Ethnicity ;Meters ;Free Verse



This file don't has the body and link

In [ ]:
# save the file as CleanOutputLoveOutput.xlsx
df.to_excel('/content/gdrive/My Drive/CleanOutputLoveOutput.xlsx')

### Question 2

# Read data from CleanOutputLoveOutput.xlsx, collect information about words, sentences, paragraphs, and punctuations of poems in this data set.


**1)**
Create a new table, which has columns - PoemID, Author, LengthOne, LengthTwo, NumLine, NumPara, NumSent, NumComma

In [ ]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [ ]:
df = pd.read_csv('./CleanOutputLoveOutput.xlsx')

#df.drop('Unnamed: 0', axis = 1, inplace=True)
df.rename(columns={'Unnamed: 0': 'PoemID'}, inplace=True)

df = df.dropna()

In [ ]:
df.head()

PoemID  ...                                          file_name
0       0  ...  /content/gdrive/Shared drives/CISC6210NLPFall2...
1       1  ...  /content/gdrive/Shared drives/CISC6210NLPFall2...
2       2  ...  /content/gdrive/Shared drives/CISC6210NLPFall2...
3       3  ...  /content/gdrive/Shared drives/CISC6210NLPFall2...
4       4  ...  /content/gdrive/Shared drives/CISC6210NLPFall2...

[5 rows x 7 columns]

In [ ]:
df1 = df.copy()

LengthOne: the total number of tokens in this poem, punctuation marks are
not included.

In [ ]:
df1['LengthOne'] = df1['body'].apply(lambda x: len(nltk.word_tokenize(x)))

d.LenthTwo: the total number of tokens in this poem, punctuation marks are
included.


In [ ]:
df1['LengthTwo'] = df1['body'].apply(lambda x: len([word for word in nltk.word_tokenize(x) if word.isalnum()]))

f. NumPara: the total number of paragraph breaks.


In [ ]:
df1['NumPara'] = df1['body'].apply(lambda x: x.count('[P]'))

e.NumLine: the total number of line breaks plus paragraph breaks. 

In [ ]:
df1['NumLine'] = df1['body'].apply(lambda x: x.count('[L]')) + df1['NumPara']

g.NumSent: the total number of complete sentences (use nltk) 

In [ ]:
df1['NumSent'] = df1['body'].apply(lambda x: len(nltk.sent_tokenize(x)))

h.NumComma: the total number of commas.

In [ ]:
df1['NumComma'] = df1['LengthOne'] - df1['NumSent']

Show the statistics information about this data set, such as average, max. and min.


In [ ]:
df1.describe()

PoemID    LengthOne  ...      NumSent     NumComma
count  1995.000000  1995.000000  ...  1995.000000  1995.000000
mean   1120.049123   296.093233  ...     9.574937   286.518296
std     643.039141   428.161048  ...    12.198368   417.660943
min       0.000000     3.000000  ...     1.000000     2.000000
25%     567.500000   142.000000  ...     4.000000   135.500000
50%    1116.000000   202.000000  ...     6.000000   194.000000
75%    1673.500000   333.000000  ...    12.000000   324.000000
max    2234.000000  9819.000000  ...   219.000000  9600.000000

[8 rows x 7 columns]

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1995 entries, 0 to 2234
Data columns (total 13 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   PoemID     1995 non-null   int64 
 1   author     1995 non-null   object
 2   title      1995 non-null   object
 3   tags       1995 non-null   object
 4   body       1995 non-null   object
 5   link       1995 non-null   object
 6   file_name  1995 non-null   object
 7   LengthOne  1995 non-null   int64 
 8   LengthTwo  1995 non-null   int64 
 9   NumPara    1995 non-null   int64 
 10  NumLine    1995 non-null   int64 
 11  NumSent    1995 non-null   int64 
 12  NumComma   1995 non-null   int64 
dtypes: int64(7), object(6)
memory usage: 218.2+ KB


In [ ]:
df1 = df1[['PoemID', 'author', 'LengthOne', 'LengthTwo', 'NumPara', 'NumLine', 'NumSent', 'NumComma']]

In [ ]:
df1.to_excel('/content/gdrive/My Drive/ProcessedLoveOutputDF1.xlsx')

2) Create three new tables, each has five columns – PoemID, Author, Body, Length, and UniCount.

In [ ]:
from nltk.corpus import stopwords 
from nltk.tokenize import word_tokenize 
nltk.download('stopwords')


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

c.Body, Length, and UniCount: tokenized words in the poem, its length, and
its vocabulary size.

In [ ]:
#The first table is for tokenized words from original poems without any processing (use nltk tokenizer).

df2 = df.copy()

df2['Body'] = df2['body'].apply(lambda x: nltk.word_tokenize(x))

df2['Length'] = df2['Body'].apply(lambda x: len(x))

from collections import Counter

df2['UniCount'] = df2['Body'].apply(lambda x: len(Counter(x)))

df2 = df2[['PoemID', 'author', 'Body', 'Length', 'UniCount']]

In [ ]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1995 entries, 0 to 2234
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   PoemID    1995 non-null   int64 
 1   author    1995 non-null   object
 2   Body      1995 non-null   object
 3   Length    1995 non-null   int64 
 4   UniCount  1995 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 93.5+ KB


In [ ]:
df2.to_excel('/content/gdrive/My Drive/ProcessedLoveOutputDF2.xlsx')

In [ ]:
#The second table is for tokenized words after stop words removal. (use nltk English stopwords list). 
df3 = df.copy()

stop_words = set(stopwords.words('english')) 

df3['Body'] = df3['body'].apply(lambda x: [w for w in nltk.word_tokenize(x) if not w in stop_words] )

df3['Length'] = df3['Body'].apply(lambda x: len(x))



In [ ]:
from collections import Counter

df3['UniCount'] = df3['Body'].apply(lambda x: len(Counter(x)))

In [ ]:
df3 = df3[['PoemID', 'author', 'Body', 'Length', 'UniCount']]

In [ ]:
df3.head()

PoemID        author  ... Length  UniCount
0       0    Nomi Stone  ...     56        36
1       1  Philip Gross  ...    159        71
2       2  A. F. Moritz  ...    326       158
3       3  R. S. Thomas  ...    168        84
4       4    Lynda Hull  ...     96        51

[5 rows x 5 columns]

In [ ]:
df3.to_excel('/content/gdrive/My Drive/ProcessedLoveOutputDF3.xlsx')

In [ ]:
#  The third table is for tokenized words after stop words removal and stemming. (use one stemmer from nltk)
from nltk.stem import PorterStemmer 
from nltk.tokenize import word_tokenize 
from nltk.corpus import stopwords 
from nltk.stem.snowball import SnowballStemmer

def remove_stemming(x):
  stop_words = set(stopwords.words('english')) 
  stemmer = SnowballStemmer("english")
  word_tokens = word_tokenize(x) 
  filtered_sentence = [w for w in word_tokens if not w in stop_words]
  stemming = []
  for w in filtered_sentence:
    stemming.append(stemmer.stem(w))
  return stemming

In [ ]:
df4 = df.copy()

In [ ]:
df4['Body'] = df4['body'].apply(lambda x: remove_stemming(x))

df4['Length'] = df4['Body'].apply(lambda x: len(x))

df4['UniCount'] = df4['Body'].apply(lambda x: len(Counter(x)))

df4 = df4[['PoemID', 'author', 'Body', 'Length', 'UniCount']]

In [ ]:
df4.head()

PoemID        author  ... Length  UniCount
0       0    Nomi Stone  ...     56        36
1       1  Philip Gross  ...    159        71
2       2  A. F. Moritz  ...    326       156
3       3  R. S. Thomas  ...    168        82
4       4    Lynda Hull  ...     96        50

[5 rows x 5 columns]

In [ ]:
df4.to_excel('/content/gdrive/My Drive/ProcessedLoveOutputDF4.xlsx')